### ISSUE TITLE
gocryptfs support to specify the volume name?

### ISSUE THREAD

**Issue by corelanc0d3r** [rfjakob/gocryptfs#557]
Hi,

I was wondering if there is a way to specify the desired volume name when mounting a gocryptfs folder?
Assuming gocryptfs was accepting similar syntax as encfs, I tried to pass the "volname" flag via -o ... but then I get this:

```
gocryptfs /path/to/cipherfolder /path/to/mountfolder -o volname="blah"
flag provided but not defined: -volname
Invalid command line: gocryptfs -volname=blah /path/to/cipherfolder /path/to/mountfolder. Try 'gocryptfs -help'.
```

What is the correct syntax to use?

thanks !

**Comment by rfjakob** [comment]
The flag is called `fsname`, and it needs to be passed to the kernel, so you have to use `-ko` (kernel option). This works:

    gocryptfs -ko fsname=blah cipher mnt

**Comment by corelanc0d3r** [comment]
Unfortunately it didn't seem to work for me:

```
corelanc0d3r@/Users/corelanc0d3r/~ % gocryptfs -ko fsname="blah" CTest /Users/corelanc0d3r/mnt/test 
Password: 
Decrypting master key
Filesystem mounted and ready.
corelanc0d3r@/Users/corelanc0d3r/ ~ % 
```
Volume was mounted and shows up as "test" (name of the mount folder), not as "blah"

**Comment by rfjakob** [comment]
Wow, looks like this is called "fsname" on Linux and "volname" on MacOS (as you said!). I only tested on Linux.

Can you try again with

gocryptfs -ko volname="blah"

?

**Comment by corelanc0d3r** [comment]
Unfortunately that didn't work either

I can confirm however that it works when mounting volumes using encfs, but perhaps encfs is using macFuse, and gocryptfs is using osxfuse?
please don't hesitate to let me know what I can do to help troubleshooting the issue

thanks !

**Comment by mike-kfed** [comment]
I can confirm that volname doesn't have consequences on macOS 11.3.1 using macfuse 4.1.2 and gocryptfs built from current master (using the `build-without-openssl.bash` script). I tried to find where it fails in the code but I couldn't explain initially what is failing -  gocryptfs autosets the volname to the basename of the mountpath, so no need to use `-ko volname=custom`, in fact that does pass volname twice to macfuse later. So I investigated further into go-fuse which looks fine too, then I compared with encfs mount options and found it (It works fine using encfs which uses libfuse directly): marking the volume "local" makes it show up in Finder with the volname. 

running `./gocryptfs -ko local cipher plain` therefore helps. Maybe the "local" flag should be default in gocryptfs?

**Comment by knaccc** [comment]
@mike-kfed I agree - I spent half an hour experimenting before I found your solution to make the gocryptfs mount appear in locations within Finder.

**Comment by rfjakob** [comment]
If the behavoir is better, we can set it per default. Could you summarize how things look with and without "-local"?

**Comment by mike-kfed** [comment]
I am not 100% certain using the "local" option is the right fix, as the icon of the drive looks different than an encfs mount - leading me to assume encfs actually does not use this flag. However with golangs  macfuse implementation it at least shows some icon in finder and so far it worked flawlessly for me - a hack worth considering? Somebody knowing macfuse and macos in depth would be better suited to correctly fix this issue I guess.

sorry for the confusion, when it finally worked I was so happy I posted the solution above, but looking into this further I realised I might be wrongly using this flag.

**Comment by knaccc** [comment]
@rfjakob In macOS Big Sur 11.4, when I mount with the "-ko local" flag, I see the mount automatically appear in the "locations" section of the left pane in every Finder window.

Without that flag, nothing appears in the "locations" section (even if the mount exists in the /Volumes folder). I think this behavior changed recently with macOS, since I think remember seeing sshfs mounts automatically appear in this section until recently, and now sshfs mounts don't show automatically.

It's not a big deal if it's a default or not, but it would be very useful to document this flag for macOS users.

**Comment by JokerQyou** [comment]
The `-ko volname=XXX` does not work, because gocryptfs sets `volname` to directory base name of the mountpoint first, then expect to override it with duplicate option-value pair. Relevant codes:

https://github.com/rfjakob/gocryptfs/blob/f06f27e7bc098e334024c365004f9303e79997d9/mount.go#L437-L442

https://github.com/rfjakob/gocryptfs/blob/f06f27e7bc098e334024c365004f9303e79997d9/mount.go#L467-L473

Basically, it would produce `-o volname=BaseNameOfMountPoint,volname=CustomVolName` option for macFUSE.
This appears logical at first glance, but it doesn't work as expected. macFUSE only reads the initial `volname` value set by gocryptfs and does not check for updates to this option later on.

I swapped these two code blocks and test on macOS Sonoma and it works as expected, the `volname` option can now be overridden correctly.

**Comment by JokerQyou** [comment]
Since macFUSE is not open source, we may never know the actual logic of its internal option parsing. But I think it's worth rewriting this part of gocryptfs so it only sets each option exactly once - use only user-provided option if possible, only inject stock options if they are not provided.
